### ISSUE TITLE
Deadlock/hang when configuring >1 consul_sd server_groups with tags

### ISSUE THREAD

**Issue by frebib** [jacksontj/promxy#313]
This is a strange one, it took me a while to pin down exactly what was going on. At first I thought the server was messing me around. It appears that whilst trying to enumerate the server_groups at first startup, promxy locks up and never resumes. Happens both with 0.0.58 and 0.0.59. Running on Linux x86_64. I haven't tried any other builds
Here is some config to reproduce the hang:
```
global:
  evaluation_interval: 60s
  scrape_interval: 60s
  scrape_timeout: 20s
promxy:
  server_groups:
  - consul_sd_configs:
    - server: localhost:8500
      datacenter: sfo07
      services:
      - prometheus-rack
      tags:
      - sw00.sfo07
  - consul_sd_configs:
    - server: localhost:8500
      datacenter: sfo07
      services:
      - prometheus-rack
      tags:
      - sw01.sfo07
```
Removing the tags from either allows promxy to start, but as soon as both have a list of tags it hangs.
Some logs, but they're not particularly useful: https://gist.github.com/frebib/00fd781a1ca921f00b49edb363252c51

Our use case is two Prometheis per rack in a server_group, with n racks, all aggregated by promxy. It's orchestrated by nomad, hence the use of consul. It's pretty early testing at the moment, so hopefully we can get this working. Cardinality and WAL size is starting to get crazy.
I'm happy to debug given that we already have the environment set up, but I'll need some pointers on where to look. I've never tried to debug deadlocks in Go before.

One other strange thing that I couldn't work out too was promxy was returning an error when one of the two instances in a group was down, but I was under the impression that it should continue working fine- that's for another issue though. It's very difficult to actually tell which instances in each server_group are up or not so some more debug logging or metrics to expose that would be great

**Comment by jacksontj** [comment]
First off, thanks for the report! I have noticed some unresponsiveness in startup, but haven't run into a deadlock (and I actually use consul a LOT) so definitely curious to track this one down.

Promxy actually has the pprof plugin () enabled, so if you have it running you can go to `http://localhost:8082/debug/pprof/` on the promxy instance to get insight into whats happening. The easiest to look at is the goroutine dump which will include what every goroutine is doing (which should make it trivial to find a deadlock). So if you can't figure it out from that data, I'd suggest grabbing a few of those goroutine dumps spread out over a couple of minutes and upload them and I can take a look.

As for the failure mode, you are correct -- if there are 2 servers in the ServerGroup and one is alive promxy should continue along (if promxy is unable to get data for a servergroup entirely it will by default error the entire response).

**Comment by jacksontj** [comment]
I can reproduce the symptoms you are seeing -- although I'm not certain that the cause is the same exactly. I just started a local consul agent in `dev` mode with the config you provided. After doing so I get the following log lines:


```
{"caller":"consul.go:360","discovery":"consul","err":"Unexpected response code: 500 (No path to datacenter)","level":"error","msg":"Error refreshing service list","ts":"2020-05-30T05:24:10.639Z"}
{"caller":"consul.go:360","discovery":"consul","err":"Unexpected response code: 500 (No path to datacenter)","level":"error","msg":"Error refreshing service list","ts":"2020-05-30T05:24:10.639Z"}
```

And after hacking in a pprof endpoint that starts at the top of main I can see that this is because the ServerGroup load hasn't completed. The initial ApplyConfig() call will only return once the ServerGroups are done applying (or are cancelled). In this case the consul SD won't complete, so the servergroup is never marked as Ready -- so the ApplyConfig never returns. So the general case that I'm showcasing here is that if the servergroups can't load for w/e reason (usually service discovery issue) promxy won't complete startup. 

Now in your specific scenario your logs didn't include anything obvious about consul discovery issues -- but I believe it is the same scenario. If I change your config to: 

``` yaml
global:
  evaluation_interval: 60s
  scrape_interval: 60s
  scrape_timeout: 20s
promxy:
  server_groups:
  - consul_sd_configs:
    - server: localhost:8500
      datacenter: dc1
      services:
      - prometheus-rack
      tags:
      - sw00.sfo07
  - consul_sd_configs:
    - server: localhost:8500
      datacenter: dc1
      services:
      - prometheus-rack
      tags:
      - sw01.sfo07
```

 (just change datacenter to dc1 which by default exists) I now have a config which doesn't match any hosts -- and in this scenario promxy just sits there on an empty terminal not starting properly. So I believe the issue is you have a servergroup (sw01.sfo07 based on the logs) which isn't returning any hosts -- so the servergroup isn't loading properly. So to unblock the issue it seems that you just don't have targets in the consul group defined there.

The secondary issue here are is how to log which servergroups are stuck and #314 will take care of that :)

**Comment by frebib** [comment]
Thanks for the quick digging!
So I came back to look at this today and without changing anything at all, it is suddenly working. Presumably over the weekend the Nomad prometheus jobs were rescheduled and now all groups are "ready".

I still have no way of telling which groups have members (how many, and which nodes) either through the logs, or from the webui (from metrics?). I noticed there don't appear to be any metrics for specific server-groups exposed by Promxy- is that by design due to the potential high cardinality? Could there be an option to expose these?

I'm also concerned about the fact that I couldn't kill the program when waiting for servergroups to be ready, so perhaps it would be good to pass the background context through to the `Ready` function in case it hangs again like that.
I'm happy to send a PR for that, I just want to get your thought on it first.

**Comment by jacksontj** [comment]
> I still have no way of telling which groups have members (how many, and which nodes) either through the logs, or from the webui (from metrics?). I noticed there don't appear to be any metrics for specific server-groups exposed by Promxy- is that by design due to the potential high cardinality? Could there be an option to expose these?

The concern here is 2 fold; (1) high cardinality -- as you suggest we could add this with options/flags to control those and (2) how to identify the various servergroups (best I can do today is by ordinal :/ ). If we think those are acceptable we can make a separate issue to add those :)

> I'm also concerned about the fact that I couldn't kill the program when waiting for servergroups to be ready, so perhaps it would be good to pass the background context through to the Ready function in case it hangs again like that.

I'd be okay with a PR for this. It sounds like the load should be moved into a background goroutine that blocks etc. so that we can have the main goroutine get to the signal listener sooner (such that it can cancel/exit). Defintely feel free to open a PR for that :)

**Comment by frebib** [comment]
> how to identify the various servergroups

There are a couple of options I can see for this
* Use some unique value/set of values from the server-group definition (may be difficult to always find something unique)
* Explicitly name the groups, and default to the ordinal if no name is specified

For groups that use service discovery, there may be some unique data, like my trivial-ish example from the top of this issue, but I imagine that's more the exception than the rule

>  It sounds like the load should be moved into a background goroutine that blocks etc.

I thought about this and I couldn't really see why it's necessary to wait for the servergroups to be ready before starting the main process. Signal handling and the http server should probably be started immediately even if the servergroups are not ready, no? Perhaps there is a reason for it that I am missing

**Comment by jacksontj** [comment]
The only reason was ease of implementation, as now to move the signal
handling outside we'll have to background something and coordinate it in
main.go
